### PR TITLE
Fix #60568 & upgrade to V2 API

### LINKED ISSUE #60568: Salt cloud vultr integration is broken, and some of the cloud utilities function make it difficult to debug
**Description**

**Setup**
https://github.com/fake-name/AutoTriever/blob/master/marshaller/salt_runner.py

**Steps to Reproduce the behavior**
Attempt to deploy a Vultr VM.
Deploy fails.

**Expected behavior**
I expect to actually create the VM.

**Screenshots**
```
Main.VpsHerder.MainThread - INFO - Creating instance...
Main.VpsHerder.MainThread - INFO -      Client name: 'test-1'
Main.VpsHerder.MainThread - INFO -      using provider: 'vultr'
Main.VpsHerder.MainThread - INFO -      kwargs: '{'image': 270, 'private_networking': False, 'size': '201', 'location': 22, 'script': '/home/client/AutoTriever/marshaller/bootstrap-salt-delay.sh'}'
salt.utils.entrypoints.MainThread - DEBUG - Using pkg_resources to load entry points
salt.utils.entrypoints.MainThread - DEBUG - Using pkg_resources to load entry points
salt.utils.lazy.MainThread - DEBUG - Could not LazyLoad parallels.avail_sizes: 'parallels' __virtual__ returned False
salt.utils.lazy.MainThread - DEBUG - LazyLoaded parallels.avail_locations
salt.utils.lazy.MainThread - DEBUG - LazyLoaded proxmox.avail_sizes
salt.cloud.MainThread - DEBUG - Generating minion keys for 'test-1'
salt.utils.lazy.MainThread - DEBUG - LazyLoaded cloud.fire_event
salt.utils.event.MainThread - DEBUG - MasterEvent PUB socket URI: /var/run/salt/master/master_event_pub.ipc
salt.utils.event.MainThread - DEBUG - MasterEvent PULL socket URI: /var/run/salt/master/master_event_pull.ipc
salt.utils.event.MainThread - DEBUG - Sending event: tag = salt/cloud/test-1/creating; data = {'provider': 'vultr:vultr', 'driver': 'vultr:vultr', 'name': 'test-1', 'profile': None, 'event': 'starting create', '_stamp': '2021-07-19T07:00:32.016509'}
salt.transport.ipc.MainThread - DEBUG - Closing IPCMessageClient instance
salt.utils.lazy.MainThread - DEBUG - LazyLoaded http.query
salt.loaded.int.utils.http.MainThread - DEBUG - Requesting URL https://api.vultr.com/v1/regions/list?api_key=XXXXXXXXXX using GET method
salt.loaded.int.utils.http.MainThread - DEBUG - Using backend: tornado
salt.loaded.int.utils.http.MainThread - DEBUG - Response Status Code: 200
salt.loaded.int.utils.http.MainThread - DEBUG - Requesting URL https://api.vultr.com/v1/os/list?api_key=XXXXXXXXXX using GET method
salt.loaded.int.utils.http.MainThread - DEBUG - Using backend: tornado
salt.loaded.int.utils.http.MainThread - DEBUG - Response Status Code: 200
salt.loaded.int.utils.http.MainThread - DEBUG - Requesting URL https://api.vultr.com/v1/plans/list?api_key=XXXXXXXXXX using GET method
salt.loaded.int.utils.http.MainThread - DEBUG - Using backend: tornado
salt.loaded.int.utils.http.MainThread - DEBUG - Response Status Code: 200
salt.loaded.int.clouds.vultrpy.MainThread - INFO - Creating Cloud VM test-1
salt.utils.event.MainThread - DEBUG - MasterEvent PUB socket URI: /var/run/salt/master/master_event_pub.ipc
salt.utils.event.MainThread - DEBUG - MasterEvent PULL socket URI: /var/run/salt/master/master_event_pull.ipc
salt.utils.event.MainThread - DEBUG - Sending event: tag = salt/cloud/test-1/requesting; data = {'kwargs': {'hostname': 'test-1', 'VPSPLANID': '201', 'enable_private_network': 'no', 'DCID': '22', 'OSID': 270, 'label': 'test-1'}, 'event': 'requesting instance', '_stamp': '2021-07-19T07:00:33.074577'}
salt.transport.ipc.MainThread - DEBUG - Closing IPCMessageClient instance
salt.loaded.int.utils.http.MainThread - DEBUG - Requesting URL https://api.vultr.com/v1/server/create?api_key=XXXXXXXXXX using POST method
salt.loaded.int.utils.http.MainThread - DEBUG - Using backend: tornado
salt.loaded.int.utils.http.MainThread - DEBUG - Response Status Code: 200
salt.loaded.int.utils.cloud.MainThread - DEBUG - Attempting function <function create.<locals>.wait_for_hostname at 0x7f998b660e50>
salt.loaded.int.utils.cloud.MainThread - DEBUG - Caught exception in wait_for_fun: '__utils__'
salt.loaded.int.utils.cloud.MainThread - DEBUG -     Traceback (most recent call last):
salt.loaded.int.utils.cloud.MainThread - DEBUG -       File "/usr/lib/python3/dist-packages/salt/utils/cloud.py", line 682, in wait_for_fun
salt.loaded.int.utils.cloud.MainThread - DEBUG -         response = fun(**kwargs)
salt.loaded.int.utils.cloud.MainThread - DEBUG -       File "/usr/lib/python3/dist-packages/salt/cloud/clouds/vultrpy.py", line 497, in wait_for_hostname
salt.loaded.int.utils.cloud.MainThread - DEBUG -         data = show_instance(vm_["name"], call="action")
salt.loaded.int.utils.cloud.MainThread - DEBUG -       File "/usr/lib/python3/dist-packages/salt/cloud/clouds/vultrpy.py", line 315, in show_instance
salt.loaded.int.utils.cloud.MainThread - DEBUG -         nodes = list_nodes_full()
salt.loaded.int.utils.cloud.MainThread - DEBUG -       File "/usr/lib/python3/dist-packages/salt/cloud/clouds/vultrpy.py", line 246, in list_nodes_full
salt.loaded.int.utils.cloud.MainThread - DEBUG -         nodes = _query("server/list")
salt.loaded.int.utils.cloud.MainThread - DEBUG -       File "/usr/lib/python3/dist-packages/salt/cloud/clouds/vultrpy.py", line 617, in _query
salt.loaded.int.utils.cloud.MainThread - DEBUG -         result = __utils__["http.query"](
salt.loaded.int.utils.cloud.MainThread - DEBUG -       File "/usr/lib/python3/dist-packages/salt/loader_context.py", line 78, in __getitem__
salt.loaded.int.utils.cloud.MainThread - DEBUG -         return self.value()[item]
salt.loaded.int.utils.cloud.MainThread - DEBUG -       File "/usr/lib/python3/dist-packages/salt/loader_context.py", line 72, in value
salt.loaded.int.utils.cloud.MainThread - DEBUG -         return loader.pack[self.name]
salt.loaded.int.utils.cloud.MainThread - DEBUG -     KeyError: '__utils__'
salt.loaded.int.utils.cloud.MainThread - DEBUG -
salt.loaded.int.utils.cloud.MainThread - DEBUG - Retrying function <function create.<locals>.wait_for_hostname at 0x7f998b660e50> on  (try 1)
salt.loaded.int.utils.cloud.MainThread - DEBUG - Caught exception in wait_for_fun: '__utils__'
salt.loaded.int.utils.cloud.MainThread - DEBUG -     Traceback (most recent call last):
salt.loaded.int.utils.cloud.MainThread - DEBUG -       File "/usr/lib/python3/dist-packages/salt/utils/cloud.py", line 682, in wait_for_fun
salt.loaded.int.utils.cloud.MainThread - DEBUG -         response = fun(**kwargs)
salt.loaded.int.utils.cloud.MainThread - DEBUG -       File "/usr/lib/python3/dist-packages/salt/cloud/clouds/vultrpy.py", line 497, in wait_for_hostname
salt.loaded.int.utils.cloud.MainThread - DEBUG -         data = show_instance(vm_["name"], call="action")
salt.loaded.int.utils.cloud.MainThread - DEBUG -       File "/usr/lib/python3/dist-packages/salt/cloud/clouds/vultrpy.py", line 315, in show_instance
salt.loaded.int.utils.cloud.MainThread - DEBUG -         nodes = list_nodes_full()
salt.loaded.int.utils.cloud.MainThread - DEBUG -       File "/usr/lib/python3/dist-packages/salt/cloud/clouds/vultrpy.py", line 246, in list_nodes_full
salt.loaded.int.utils.cloud.MainThread - DEBUG -         nodes = _query("server/list")
salt.loaded.int.utils.cloud.MainThread - DEBUG -       File "/usr/lib/python3/dist-packages/salt/cloud/clouds/vultrpy.py", line 617, in _query
salt.loaded.int.utils.cloud.MainThread - DEBUG -         result = __utils__["http.query"](
salt.loaded.int.utils.cloud.MainThread - DEBUG -       File "/usr/lib/python3/dist-packages/salt/loader_context.py", line 78, in __getitem__
salt.loaded.int.utils.cloud.MainThread - DEBUG -         return self.value()[item]
salt.loaded.int.utils.cloud.MainThread - DEBUG -       File "/usr/lib/python3/dist-packages/salt/loader_context.py", line 72, in value
salt.loaded.int.utils.cloud.MainThread - DEBUG -         return loader.pack[self.name]
salt.loaded.int.utils.cloud.MainThread - DEBUG -     KeyError: '__utils__'
salt.loaded.int.utils.cloud.MainThread - DEBUG -

<Repeat 900 (!) times>
```

Note that the above is with a patched `wait_for_fun()` implementation to actually output a useful traceback:

```python

def wait_for_fun(fun, timeout=900, **kwargs):
    """
    Wait until a function finishes, or times out
    """
    start = time.time()
    log.debug("Attempting function %s", fun)
    trycount = 0
    while True:
        trycount += 1
        try:
            response = fun(**kwargs)
            if not isinstance(response, bool):
                return response
        except Exception as exc:  # pylint: disable=broad-except
            log.debug("Caught exception in wait_for_fun: %s", exc)
            for line in traceback.format_exc().split("\n"):
                log.debug("    %s", line)
            time.sleep(1)
            log.debug("Retrying function %s on  (try %s)", fun, trycount)
        if time.time() - start > timeout:
            log.error("Function timed out: %s", timeout)
            return False
```

**Versions Report**

```
client@salt:~/AutoTriever/marshaller$ salt --versions-report
Salt Version:
          Salt: 3003.1

Dependency Versions:
          cffi: Not Installed
      cherrypy: Not Installed
      dateutil: 2.7.3
     docker-py: Not Installed
         gitdb: 2.0.6
     gitpython: 3.0.7
        Jinja2: 2.10.1
       libgit2: Not Installed
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: Not Installed
      pycrypto: 2.6.1
  pycryptodome: 3.6.1
        pygit2: Not Installed
        Python: 3.8.10 (default, Jun  2 2021, 10:49:15)
  python-gnupg: 0.4.5
        PyYAML: 5.3.1
         PyZMQ: 18.1.1
         smmap: 2.0.5
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.2

System Versions:
          dist: ubuntu 20.04 focal
        locale: utf-8
       machine: x86_64
       release: 5.4.0-77-generic
        system: Linux
       version: Ubuntu 20.04 focal
```
This is basically the current version from `https://repo.saltproject.io/py3/ubuntu/20.04/amd64/latest`

### PR DESCRIPTION
**Pull request by gizahNL**
### What does this PR do?
vultrpy is fixed so that it doesn't stacktrace in an endless loop and actually bootstraps salt on newly deployed VM's.

Furthermore the Vultr API consumed is rewritten to V2, and some missing functions are now implemented.

### What issues does this PR fix or reference?
Fixes: #60568 & previously unimplemented start/stop functions are now implemented.

### Breaking Changes:
-V2 API regions are a string instead of a numeric ID
-V2 API plans (salt: sizes) are a string instead of a numeric ID
-ssh_key_names now actually takes the key names, instead of ID, as
 suggested by documentation

### Merge requirements satisfied?
- [ ] Docs
- [ ] Changelog
- [x] Tests written/updated

### Commits signed with GPG?
No

### CODE REVIEW

**Comment by garethgreenaway** [comment]
@gizahNL Thanks for the PR.  Looking around a bit this seems like it might be an issue with pylint being overly zealous with it's checking.  This patch should allow your PR to pass:

```
diff --git a/salt/cloud/clouds/vultrpy.py b/salt/cloud/clouds/vultrpy.py
index 4181c9fc34..cdd90084cc 100644
--- a/salt/cloud/clouds/vultrpy.py
+++ b/salt/cloud/clouds/vultrpy.py
@@ -752,6 +752,7 @@ def _query(path, method="GET", data=None):
 
     data = salt.utils.json.dumps(data)
     requester = getattr(requests, method.lower())
+    # pylint: disable=not-callable
     request = requester(
         url,
         data=data,
@@ -760,6 +761,7 @@ def _query(path, method="GET", data=None):
             "Content-Type": "application/json",
         },
     )
+    # pylint: enable=not-callable
     if request.status_code > 299:
         raise SaltCloudSystemExit(
             "An error occurred while querying Vultr. HTTP Code: {}  "
```

**Comment by Ch3LL** [comment]
@gizahNL can you apply @garethgreenaway 's suggestion?

**Comment by gizahNL** [comment]
> @gizahNL can you apply @garethgreenaway 's suggestion?

Done & rebased on current master

File: salt/cloud/clouds/vultrpy.py [hunk 1]
```diff
@@ -624,28 +745,45 @@ def _query(path, method="GET", data=None, params=None, header_dict=None, decode=
...
+        raise SaltCloudSystemExit(
+            "An error occurred while querying Vultr. HTTP Code: {}  "
+            "Error: '{}'".format(
+                request.status_code,
+                # request.read()
+                request.text,
+            )
+        )
+
+    log.debug(request.url)
```
Comment (Ch3LL) [inline salt/cloud/clouds/vultrpy.py:775]: Would it be possible for this url to contain sensitive information, such as a password? If so, I don't think we should log it

Also, I think we need more information around this log. Possibly `log.debug("Querying url: {}".format(request.url))`

**Review by Ch3LL — CHANGES_REQUESTED** [review]
Thanks for doing the work to get this module ready for vultur v2. My concern is backwards compatibility. Do users that use vultr have to specify or enable v2? I worry if a user is not prepared to use the v2 api yet this will break things for them.

**Comment by Ch3LL** [comment]
Oh I also forgot, this will also require a changelog

File: salt/cloud/clouds/vultrpy.py (hunk 1 above)
Comment (gizahNL) [reply salt/cloud/clouds/vultrpy.py:775]: iirc there should be no sensitive information, I'd have to take another look at the code as it's now quite a while ago that I wrote it ;)

iirc tokens & such are only transmitted via headers, so they wouldn't be logged here. 

It's possible SSH public keys could be set as URL params (iirc they're json data posted to an API endpoint, so again should be safe).
Besides that I don't think there could be sensitive information, unless we consider machine ID's and such as sensitive.

I'll take an critical look at it when I get back to it & modify the debug statement as requested.

**Comment by gizahNL** [comment]
> Thanks for doing the work to get this module ready for vultur v2. My concern is backwards compatibility. Do users that use vultr have to specify or enable v2? I worry if a user is not prepared to use the v2 api yet this will break things for them.
I completely replaced all old V1 code, and V2 is not completely backwards compatible: as noted in the PR message identifiers for regions, sizes have changed, as well as for SSH key names (this is now as was suggested in the documentation taking the SSH key name instead of the ID).

For us (N=1) Vultr was completely unusable due to #60568 and when fixing that bug I decided to go the extra mile and rewrite for V2 to ensure future compatibility when/if Vultr decides to shut down V1.

**Comment by Ch3LL** [comment]
Will users have to change things on their side of things though? Like will they have to enable the V2 api? Or update setup on Salt side of things?

**Comment by Ch3LL** [comment]
bump @gizahNL did you see my comments above?

**Review by twangboy — CHANGES_REQUESTED** [review]
This still needs a changelog and it needs to be rebased and conflicts resolved

**Comment by RyIot3329** [comment]
Bumping this again. Wondering if this will ever get merged or if anyone has any updates on the vultr v2 api implementation.

**Comment by fake-name** [comment]
Vultr is turning off the V1 API June 30th, so this is gonna go from a "nice to have" to an emergency pretty fast.

**Pull request closed without merging**